src/evaluators.jl

```diff
@@ -0,0 +1,122 @@
+
+import Base: show
+
+# WIP: definition of AD factory
+abstract type AbstractADFactory end
+
+struct ADFactory <: AbstractADFactory
+    Jgₓ::AD.StateJacobianAD
+    Jgᵤ::AD.DesignJacobianAD
+end
+
+abstract type AbstractNLPEvaluator end
+
+struct ReducedSpaceEvaluator{T} <: AbstractNLPEvaluator
+    model::AbstractFormulation
+    x::AbstractVector{T}
+    p::AbstractVector{T}
+
+    x_min::AbstractVector{T}
+    x_max::AbstractVector{T}
+    u_min::AbstractVector{T}
+    u_max::AbstractVector{T}
+
+    ad::ADFactory
+    ε_tol::Float64
+end
+
+function ReducedSpaceEvaluator(model, x, u, p; ε_tol=1e-12)
+    jx, ju = init_ad_factory(model, x, u, p)
+    ad = ADFactory(jx, ju)
+    u_min, u_max = bounds(model, Control())
+    x_min, x_max = bounds(model, State())
+    return ReducedSpaceEvaluator(model, x, p, x_min, x_max, u_min, u_max,
+                                 ad, ε_tol)
+end
+
+n_variables(nlp::ReducedSpaceEvaluator) = length(nlp.u_min)
+# n_constraints(nlp::ReducedSpaceEvaluator) = length(nlp.u)
+
+function update!(nlp::ReducedSpaceEvaluator, u)
```

Comment:
Each time we have a new `u`, we need to recompute the corresponding `x`

Reply:
This whole module is excellent.